src/validator/OrderByValidator.cpp

```diff
@@ -13,26 +13,47 @@ namespace nebula {
-    outputs_ = inputCols();
     auto &factors = sentence->factors();
-    auto *pool = qctx_->objPool();
     for (auto &factor : factors) {
-        if (factor->expr()->kind() == Expression::Kind::kLabel) {
-            auto *label = static_cast<const LabelExpression*>(factor->expr());
-            auto *expr = InputPropertyExpression::make(pool, label->name());
-            factor->setExpr(expr);
-        }
-        if (factor->expr()->kind() != Expression::Kind::kInputProperty) {
+        if (factor->expr()->kind() == Expression::Kind::kInputProperty) {
+            auto expr = static_cast<InputPropertyExpression*>(factor->expr());
+            NG_RETURN_IF_ERROR(deduceExprType(expr));
+            NG_RETURN_IF_ERROR(deduceProps(expr, exprProps_));
+            const auto &cols = inputCols();
+            auto& name = expr->prop();
+            auto eq = [&](const ColDef& col) { return col.name == name; };
+            auto iter = std::find_if(cols.cbegin(), cols.cend(), eq);
+            size_t colIdx = std::distance(cols.cbegin(), iter);
+            colOrderTypes_.emplace_back(std::make_pair(colIdx, factor->orderType()));
+        } else if (factor->expr()->kind() == Expression::Kind::kVarProperty) {
+            auto expr = static_cast<VariablePropertyExpression*>(factor->expr());
+            NG_RETURN_IF_ERROR(deduceExprType(expr));
+            NG_RETURN_IF_ERROR(deduceProps(expr, exprProps_));
+            const auto &cols = vctx_->getVar(expr->sym());
+            auto& name = expr->prop();
+            auto eq = [&](const ColDef& col) { return col.name == name; };
+            auto iter = std::find_if(cols.cbegin(), cols.cend(), eq);
+            size_t colIdx = std::distance(cols.cbegin(), iter);
+            colOrderTypes_.emplace_back(std::make_pair(colIdx, factor->orderType()));
+        } else {
             return Status::SemanticError("Order by with invalid expression `%s'",
                                           factor->expr()->toString().c_str());
         }
-        auto expr = static_cast<InputPropertyExpression*>(factor->expr());
-        NG_RETURN_IF_ERROR(deduceExprType(expr));
-        auto& name = expr->prop();
-        auto eq = [&](const ColDef& col) { return col.name == name; };
-        auto iter = std::find_if(outputs_.cbegin(), outputs_.cend(), eq);
-        size_t colIdx = std::distance(outputs_.cbegin(), iter);
-        colOrderTypes_.emplace_back(std::make_pair(colIdx, factor->orderType()));
+    }
+
+
+    if (!exprProps_.inputProps().empty() && !exprProps_.varProps().empty()) {
+        return Status::SemanticError("Not support both input and variable.");
+    } else if (!exprProps_.inputProps().empty()) {
+        outputs_ = inputCols();
+    } else if (!exprProps_.varProps().empty()) {
+        if (!userDefinedVarNameList_.empty()) {
+            if (userDefinedVarNameList_.size() != 1) {
+                return Status::SemanticError("Multiple user defined vars not supported yet.");
```

Comment:
"Multiple user-defined vars are not supported yet."

Reply:
ok